# L2 Escape Model

Consequent Reversibility was not included in the final escape models because, in this subgroup structure, it is partially encoded by Goal_Type and produced unstable estimates when entered together with Goal_Type.

Import Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

## Overall Subgroup
This subset of data explores observations when _Focus_ = I and _Focus_= They.
1. Create escape_L2
2. Counts and Proportions
3. Logistic Models
    - E1: Agency
    - E2: Goal Type
    - E3: Additive Model
    - E4: Interactive Model 
4. Model Comparison
5. Summary Table

Read Data

In [2]:
subgroup_theoretical = pd.read_csv("../../data/processed/subgroup_theoretical.csv")

1. Create escape_2

Binary Outcome: include responses that are not "L2_other": those who managed to escape L2_other by choosing either a correct response or an L1_transfer response.

In [3]:
subgroup_theoretical["escape_L2"] = (
    subgroup_theoretical["Response_Full"] != "L2_other"
).astype(int)

2. Counts and Proportions

In [4]:
escape_L2_counts = pd.crosstab(
    subgroup_theoretical["escape_L2"],
    subgroup_theoretical["Response_Full"])

escape_L2_counts

Response_Full,L1_transfer,L2_other,correct,missing_response
escape_L2,,,,
0,0,38,0,0
1,96,0,61,3


In [5]:
escape_L2_props = subgroup_theoretical["escape_L2"].value_counts(normalize = True)
escape_L2_props

escape_L2
1    0.808081
0    0.191919
Name: proportion, dtype: float64

From 198 observations, almost 80% of responses managed to escape L2_other options. 

Counts and Proportions by Condition:

In [6]:
counts_condition = pd.crosstab(
    [subgroup_theoretical["Goal_Type"], subgroup_theoretical["Agent"]],
    subgroup_theoretical["escape_L2"],
    margins = True)

counts_condition

escape_L2                 0    1  All
Goal_Type         Agent              
goal_frequent     0      14   28   42
                  1       2   22   24
goal_non_frequent 0       9   33   42
                  1       1   23   24
no_goal           0       8   34   42
                  1       4   20   24
All                      38  160  198

In [7]:
props_condition = pd.crosstab(
    [subgroup_theoretical["Goal_Type"], subgroup_theoretical["Agent"]],
    subgroup_theoretical["escape_L2"],
    normalize= "index")

props_condition

escape_L2                       0         1
Goal_Type         Agent                    
goal_frequent     0      0.333333  0.666667
                  1      0.083333  0.916667
goal_non_frequent 0      0.214286  0.785714
                  1      0.041667  0.958333
no_goal           0      0.190476  0.809524
                  1      0.166667  0.833333

Agency appears to help participants escape L2. 

The condition with the strongest L1 transfer (goal_non_frequent, agent=1) also exhibits the highest L2 escape rate (95.8%). This suggests that failure does not occur at the L2 suppression stage, but at the subsequent resolution stage.

In addition, if there is a goal representation available, agency seems to help participants commit by leaving the L2 space; however, if there is no goal representation, agency seems to have much less to work with -which suggests that we could be dealing with an entire different behaviour when it comes to escaping l2 (and subsequent resolving) when having a goal representation available (*goal_frequent* and *goal_non_frequent*) vs. not having a goal representation available (*no_goal*).

3. Logistic Models

Import Libraries

In [8]:
import statsmodels.formula.api as smf
from scipy.stats import chi2

In [9]:
subgroup_theoretical["escape_L2"].value_counts()

escape_L2
1    160
0     38
Name: count, dtype: int64

**Model E1**

escape_L2 ~ Agent

In [10]:
E1 = smf.logit(
    "escape_L2 ~ Agent",
    data=subgroup_theoretical
    ).fit()

print(E1.summary())

Optimization terminated successfully.
         Current function value: 0.471025
         Iterations 6
                           Logit Regression Results                           
Dep. Variable:              escape_L2   No. Observations:                  198
Model:                          Logit   Df Residuals:                      196
Method:                           MLE   Df Model:                            1
Date:                Thu, 09 Jul 2026   Pseudo R-squ.:                 0.03675
Time:                        13:35:59   Log-Likelihood:                -93.263
converged:                       True   LL-Null:                       -96.821
Covariance Type:            nonrobust   LLR p-value:                  0.007642
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      1.1199      0.207      5.414      0.000       0.714       1.525
Agent          1.1086      0.

**Model E2**

escape_L2 ~ Goal_Type

In [11]:
E2 = smf.logit(
    "escape_L2 ~ Goal_Type",
    data= subgroup_theoretical
    ).fit()

print(E2.summary())

Optimization terminated successfully.
         Current function value: 0.484442
         Iterations 6
                           Logit Regression Results                           
Dep. Variable:              escape_L2   No. Observations:                  198
Model:                          Logit   Df Residuals:                      195
Method:                           MLE   Df Model:                            2
Date:                Thu, 09 Jul 2026   Pseudo R-squ.:                0.009309
Time:                        13:35:59   Log-Likelihood:                -95.920
converged:                       True   LL-Null:                       -96.821
Covariance Type:            nonrobust   LLR p-value:                    0.4060
                                     coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------------
Intercept                          1.1394      0.287      3.967     

In [12]:
E3 = smf.logit(
    "escape_L2 ~ Agent + Goal_Type",
    data= subgroup_theoretical
    ).fit()

print(E3.summary())

Optimization terminated successfully.
         Current function value: 0.466314
         Iterations 6
                           Logit Regression Results                           
Dep. Variable:              escape_L2   No. Observations:                  198
Model:                          Logit   Df Residuals:                      194
Method:                           MLE   Df Model:                            3
Date:                Thu, 09 Jul 2026   Pseudo R-squ.:                 0.04638
Time:                        13:35:59   Log-Likelihood:                -92.330
converged:                       True   LL-Null:                       -96.821
Covariance Type:            nonrobust   LLR p-value:                   0.02954
                                     coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------------
Intercept                          0.8081      0.311      2.599     

**Model E4**

escape_L2 ~ Agent * Goal_Type

In [13]:
E4 = smf.logit(
    "escape_L2 ~ Agent * Goal_Type",
    data=subgroup_theoretical
    ).fit()

print(E4.summary())

Optimization terminated successfully.
         Current function value: 0.458893
         Iterations 7
                           Logit Regression Results                           
Dep. Variable:              escape_L2   No. Observations:                  198
Model:                          Logit   Df Residuals:                      192
Method:                           MLE   Df Model:                            5
Date:                Thu, 09 Jul 2026   Pseudo R-squ.:                 0.06156
Time:                        13:35:59   Log-Likelihood:                -90.861
converged:                       True   LL-Null:                       -96.821
Covariance Type:            nonrobust   LLR p-value:                   0.03590
                                           coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------------------
Intercept                                0.6931      0.3

4. Model Comparisons

In [14]:
from scipy.stats import chi2

def lr_test(model_small, model_large):

    LR = 2 * (model_large.llf - model_small.llf)

    df = model_large.df_model - model_small.df_model

    p = chi2.sf(LR, df)

    return pd.Series({
        "LR": LR,
        "df": df,
        "p_value": p
    })

In [15]:
lr_test(E1, E3)

LR         1.865633
df         2.000000
p_value    0.393444
dtype: float64

In [16]:
lr_test(E1, E4)

LR         4.804579
df         4.000000
p_value    0.307943
dtype: float64

5. Summary Table

In [17]:
goal_behaviour = ["Combined", "Combined", "Combined", "Combined", "Combined", "Combined", "Combined", "Combined", "Combined", "Combined", "Combined"]

model_name = ["E1", "E2", "E2", "E3", "E3", "E3", "E4", "E4", "E4", "E4", "E4"]

outcome = ["Escape_L2","Escape_L2", "Escape_L2", "Escape_L2", "Escape_L2", "Escape_L2", "Escape_L2", "Escape_L2", "Escape_L2", "Escape_L2", "Escape_L2"]

subgroup = ["Overall", "Overall", "Overall", "Overall", "Overall", "Overall", "Overall", "Overall", "Overall", "Overall", "Overall"]

formula = ["escape_L2 ~ Agent", "escape_L2 ~ Goal_Type", "escape_L2 ~ Goal_Type", "escape_L2 ~ Agent + Goal_Type", "escape_L2 ~ Agent + Goal_Type", "escape_L2 ~ Agent + Goal_Type", "escape_L2 ~ Agent * Goal_Type", "escape_L2 ~ Agent * Goal_Type", "escape_L2 ~ Agent * Goal_Type", "escape_L2 ~ Agent * Goal_Type", "escape_L2 ~ Agent * Goal_Type"]

predictor= ["Agent", "goal_non_frequent", "no_goal", "goal_non_frequent", "no_goal", "Agent",  "goal_non_frequent", "no_goal", "Agent", "Agent:goal_non_frequent", "Agent:no_goal"]

coef = [1.1086, 0.5833, 0.3646, 0.6038, 0.3785, 1.1182, 0.6061, 0.7538, 1.7047, 0.1315, -1.5422]

std_error =[0.448, 0.448, 0.429, 0.455, 0.437, 0.450, 0.499, 0.511, 0.808, 1.356, 1.052]

coef_p = [0.013, 0.193, 0.396, 0.185, 0.387, 0.013, 0.224, 0.141, 0.035, 0.923, 0.143]

LLR_p = [0.007642, 0.4060, 0.4060, 0.02954, 0.02954, 0.02954, 0.03590, 0.03590, 0.03590, 0.03590, 0.03590]

observations= [198, 198, 198, 198, 198, 198, 198, 198, 198, 198, 198]

Takeaway = []

In [18]:
escape_summary_overall = pd.DataFrame({
    "Goal_Behaviour": goal_behaviour,
    "Model": model_name,
    "Outcome": outcome,
    "Subgroup": subgroup,
    "Formula": formula, 
    "Predictor": predictor,
    "Coef": coef,
    "Std_err": std_error,
    "Coef_p": coef_p,
    "Model_LLR_p": LLR_p,
    "N": observations
})

escape_summary_overall

,Goal_Behaviour,Model,Outcome,Subgroup,Formula,Predictor,Coef,Std_err,Coef_p,Model_LLR_p,N
0,Combined,E1,Escape_L2,Overall,escape_L2 ~ Agent,Agent,1.1086,0.448,0.013,0.007642,198
1,Combined,E2,Escape_L2,Overall,escape_L2 ~ Goal_Type,goal_non_frequent,0.5833,0.448,0.193,0.406000,198
2,Combined,E2,Escape_L2,Overall,escape_L2 ~ Goal_Type,no_goal,0.3646,0.429,0.396,0.406000,198
3,Combined,E3,Escape_L2,Overall,escape_L2 ~ Agent + Goal_Type,goal_non_frequent,0.6038,0.455,0.185,0.029540,198
4,Combined,E3,Escape_L2,Overall,escape_L2 ~ Agent + Goal_Type,no_goal,0.3785,0.437,0.387,0.029540,198
5,Combined,E3,Escape_L2,Overall,escape_L2 ~ Agent + Goal_Type,Agent,1.1182,0.450,0.013,0.029540,198
6,Combined,E4,Escape_L2,Overall,escape_L2 ~ Agent * Goal_Type,goal_non_frequent,0.6061,0.499,0.224,0.035900,198
7,Combined,E4,Escape_L2,Overall,escape_L2 ~ Agent * Goal_Type,no_goal,0.7538,0.511,0.141,0.035900,198
8,Combined,E4,Escape_L2,Overall,escape_L2 ~ Agent * Goal_Type,Agent,1.7047,0.808,0.035,0.035900,198
9,Combined,E4,Escape_L2,Overall,escape_L2 ~ Agent * Goal_Type,Agent:goal_non_frequent,0.1315,1.356,0.923,0.035900,198


## Actors Subgroup
This subset of data explores observations only when _Focus_ = I.
1. Create escape_L2
2. Counts and Proportions
3. Logistic Models
    - E1_A: Agency
    - E2_A: Goal Type
    - E3_A: Additive Model
    - E4_A: Interactive Model 
4. Model Comparison
5. Summary Table

Read Data

In [19]:
subgroup_actors = subgroup_theoretical[subgroup_theoretical["Focus"] == "I"].copy()

1. Create escape_L2

Binary Outcome: within actors, include responses that are not *L2_other*; that is, those who managed to escape *L2_other* by choosing either a *correct* response or an *L1_transfer* response.

In [20]:
subgroup_actors["escape_L2"] = (
    subgroup_actors["Response_Full"] != "L2_other"
).astype(int)

2. Counts and Proportions

In [21]:
escape_L2_counts = pd.crosstab(
    subgroup_actors["escape_L2"],
    subgroup_actors["Response_Full"])

escape_L2_counts

Response_Full,L1_transfer,L2_other,correct,missing_response
escape_L2,,,,
0,0,24,0,0
1,53,0,40,3


In [22]:
escape_L2_props = subgroup_actors["escape_L2"].value_counts(normalize = True)
escape_L2_props

escape_L2
1    0.8
0    0.2
Name: proportion, dtype: float64

From 120 actors' observations (from the original 198), almost 80% of responses managed to escape *L2_other* options just as with the overall subgroup.

Counts and Proportions by Condition:

In [23]:
counts_condition = pd.crosstab(
    [subgroup_actors["Goal_Type"], subgroup_actors["Agent"]],
    subgroup_actors["escape_L2"],
    margins = True)

counts_condition

escape_L2                 0   1  All
Goal_Type         Agent             
goal_frequent     0       8  16   24
                  1       2  14   16
goal_non_frequent 0       8  16   24
                  1       1  15   16
no_goal           0       3  21   24
                  1       2  14   16
All                      24  96  120

In [24]:
props_condition = pd.crosstab(
    [subgroup_actors["Goal_Type"], subgroup_actors["Agent"]],
    subgroup_actors["escape_L2"],
    normalize= "index")

props_condition

escape_L2                       0         1
Goal_Type         Agent                    
goal_frequent     0      0.333333  0.666667
                  1      0.125000  0.875000
goal_non_frequent 0      0.333333  0.666667
                  1      0.062500  0.937500
no_goal           0      0.125000  0.875000
                  1      0.125000  0.875000

Just as with the Overall subgroup, within actors, Agency also appears to help participants escape L2. 

The condition with the strongest L1 transfer (goal_non_frequent, agent=1) keeps exhibiting the highest L2 escape rate (93.7%, slightly lower than in the overall subgroup (95.8%)). Still, this keeps hinting at  failure not occuring at the L2 suppression stage, but at the subsequent resolution stage.

Considering differences regarding goal representation availability, these proportions support that, if there is a goal representation available, agency seems to help participants commit by leaving the L2 space; however, if there is no goal representation, agency has nothing to work with by showing the exact same proportions. 

This further supports the idea that we could be dealing with an entire different behaviour when it comes to escaping l2 (and subsequent resolving) when having a goal representation available (*goal_frequent* and *goal_non_frequent*) vs. not having a goal representation available (*no_goal*) that should be explored later on. 

3. Logistic Models

In [25]:
subgroup_actors["escape_L2"].value_counts()

escape_L2
1    96
0    24
Name: count, dtype: int64

**E1_A Model** 

escape_L2 ~ Agent

In [26]:
E1_A = smf.logit(
    "escape_L2 ~ Agent",
    data=subgroup_actors
    ).fit()

print(E1_A.summary())

Optimization terminated successfully.
         Current function value: 0.479908
         Iterations 6
                           Logit Regression Results                           
Dep. Variable:              escape_L2   No. Observations:                  120
Model:                          Logit   Df Residuals:                      118
Method:                           MLE   Df Model:                            1
Date:                Thu, 09 Jul 2026   Pseudo R-squ.:                 0.04096
Time:                        13:35:59   Log-Likelihood:                -57.589
converged:                       True   LL-Null:                       -60.048
Covariance Type:            nonrobust   LLR p-value:                   0.02657
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      1.0259      0.267      3.836      0.000       0.502       1.550
Agent          1.1259      0.

**E2_A Model** 

escape_L2 ~  Goal_Type

In [27]:
E2_A = smf.logit(
    "escape_L2 ~ Goal_Type",
    data=subgroup_actors
    ).fit()

print(E2_A.summary())

Optimization terminated successfully.
         Current function value: 0.490756
         Iterations 6
                           Logit Regression Results                           
Dep. Variable:              escape_L2   No. Observations:                  120
Model:                          Logit   Df Residuals:                      117
Method:                           MLE   Df Model:                            2
Date:                Thu, 09 Jul 2026   Pseudo R-squ.:                 0.01928
Time:                        13:35:59   Log-Likelihood:                -58.891
converged:                       True   LL-Null:                       -60.048
Covariance Type:            nonrobust   LLR p-value:                    0.3143
                                     coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------------
Intercept                          1.0986      0.365      3.009     

**E3_A Model**

escape_L2 ~  Agent + Goal_Type

In [28]:
E3_A = smf.logit(
    "escape_L2 ~ Agent + Goal_Type",
    data= subgroup_actors
    ).fit()

print(E3_A.summary())

Optimization terminated successfully.
         Current function value: 0.469898
         Iterations 6
                           Logit Regression Results                           
Dep. Variable:              escape_L2   No. Observations:                  120
Model:                          Logit   Df Residuals:                      116
Method:                           MLE   Df Model:                            3
Date:                Thu, 09 Jul 2026   Pseudo R-squ.:                 0.06096
Time:                        13:35:59   Log-Likelihood:                -56.388
converged:                       True   LL-Null:                       -60.048
Covariance Type:            nonrobust   LLR p-value:                   0.06234
                                     coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------------
Intercept                          0.7203      0.400      1.799     

**E4_A Model**

escape_L2 ~ Agent * Goal_Type

In [29]:
E4_A = smf.logit(
    "escape_L2 ~ Agent * Goal_Type",
    data= subgroup_actors
    ).fit()

print(E4_A.summary())

Optimization terminated successfully.
         Current function value: 0.461604
         Iterations 7
                           Logit Regression Results                           
Dep. Variable:              escape_L2   No. Observations:                  120
Model:                          Logit   Df Residuals:                      114
Method:                           MLE   Df Model:                            5
Date:                Thu, 09 Jul 2026   Pseudo R-squ.:                 0.07753
Time:                        13:35:59   Log-Likelihood:                -55.392
converged:                       True   LL-Null:                       -60.048
Covariance Type:            nonrobust   LLR p-value:                   0.09726
                                           coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------------------
Intercept                                0.6931      0.4

4. Model comparisons

In [30]:
lr_test(E1_A, E3_A)

LR         2.402544
df         2.000000
p_value    0.300811
dtype: float64

In [31]:
lr_test(E1_A, E4_A)

LR         4.393054
df         4.000000
p_value    0.355418
dtype: float64

5. Summary Table

escape_summary_actors

In [32]:
goal_behaviour = ["Combined", "Combined", "Combined", "Combined", "Combined", "Combined", "Combined", "Combined", "Combined", "Combined", "Combined"]

model_name = ["E1_A", "E2_A", "E2_A", "E3_A", "E3_A", "E3_A", "E4_A", "E4_A", "E4_A", "E4_A", "E4_A"]

outcome = ["Escape_L2", "Escape_L2", "Escape_L2", "Escape_L2", "Escape_L2", "Escape_L2", "Escape_L2", "Escape_L2", "Escape_L2", "Escape_L2", "Escape_L2"]

subgroup = ["Actors", "Actors", "Actors", "Actors", "Actors", "Actors", "Actors", "Actors", "Actors", "Actors", "Actors"]

formula = ["escape_L2 ~ Agent", "escape_L2 ~ Goal_Type", "escape_L2 ~ Goal_Type", "escape_L2 ~ Agent + Goal_Type", "escape_L2 ~ Agent + Goal_Type", "escape_L2 ~ Agent + Goal_Type", "escape_L2 ~ Agent * Goal_Type", "escape_L2 ~ Agent * Goal_Type", "escape_L2 ~ Agent * Goal_Type", "escape_L2 ~ Agent * Goal_Type", "escape_L2 ~ Agent * Goal_Type"]

predictor= ["Agent", "goal_non_frequent", "no_goal", "goal_non_frequent", "no_goal", "Agent",  "goal_non_frequent", "no_goal", "Agent", "Agent:goal_non_frequent", "Agent:no_goal"]

coef = [1.1259, 0.1382, 0.8473, 0.1448, 0.8790, 1.1455, -3.881e-16, 1.2528, 1.2528, 0.7621, -1.2528]

std_error= [0.543, 0.526, 0.602, 0.538, 0.613, 0.547, 0.612, 0.754, 0.871, 1.419, 1.308]

coef_p = [0.038, 0.793, 0.159, 0.788, 0.151, 0.036, 1.000, 0.097, 0.150, 0.591, 0.338]

LLR_p = [0.02657, 0.3143, 0.3143, 0.06234, 0.06234, 0.06234, 0.09726, 0.09726, 0.09726, 0.09726, 0.09726]

observations = [120, 120, 120, 120, 120, 120, 120, 120, 120, 120, 120]

Takeaway = []

In [33]:
escape_summary_actors = pd.DataFrame({
    "Goal_Behaviour": goal_behaviour,
    "Model": model_name,
    "Outcome": outcome,
    "Subgroup": subgroup,
    "Formula": formula, 
    "Predictor": predictor,
    "Coef": coef,
    "Std_err": std_error,
    "Coef_p": coef_p,
    "Model_LLR_p": LLR_p,
    "N": observations
})

escape_summary_actors

,Goal_Behaviour,Model,Outcome,Subgroup,Formula,Predictor,Coef,Std_err,Coef_p,Model_LLR_p,N
0,Combined,E1_A,Escape_L2,Actors,escape_L2 ~ Agent,Agent,1.125900e+00,0.543,0.038,0.02657,120
1,Combined,E2_A,Escape_L2,Actors,escape_L2 ~ Goal_Type,goal_non_frequent,1.382000e-01,0.526,0.793,0.31430,120
2,Combined,E2_A,Escape_L2,Actors,escape_L2 ~ Goal_Type,no_goal,8.473000e-01,0.602,0.159,0.31430,120
3,Combined,E3_A,Escape_L2,Actors,escape_L2 ~ Agent + Goal_Type,goal_non_frequent,1.448000e-01,0.538,0.788,0.06234,120
4,Combined,E3_A,Escape_L2,Actors,escape_L2 ~ Agent + Goal_Type,no_goal,8.790000e-01,0.613,0.151,0.06234,120
5,Combined,E3_A,Escape_L2,Actors,escape_L2 ~ Agent + Goal_Type,Agent,1.145500e+00,0.547,0.036,0.06234,120
6,Combined,E4_A,Escape_L2,Actors,escape_L2 ~ Agent * Goal_Type,goal_non_frequent,-3.881000e-16,0.612,1.000,0.09726,120
7,Combined,E4_A,Escape_L2,Actors,escape_L2 ~ Agent * Goal_Type,no_goal,1.252800e+00,0.754,0.097,0.09726,120
8,Combined,E4_A,Escape_L2,Actors,escape_L2 ~ Agent * Goal_Type,Agent,1.252800e+00,0.871,0.150,0.09726,120
9,Combined,E4_A,Escape_L2,Actors,escape_L2 ~ Agent * Goal_Type,Agent:goal_non_frequent,7.621000e-01,1.419,0.591,0.09726,120
